### PR TITLE
fix(auth): storage not synced before accessing when the app boots from killed state

### PR DESCRIPTION
**Pull request by Samaritan1011001**
#### Description of changes

**[V5 only bug fix]**
We sync the storage before it's read so when the LastAuthUser is accessed it's reading the latest values. 

#### Issue #, if available

https://github.com/aws-amplify/amplify-js/issues/8384

#### Description of how you validated changes
1. Linked the changed library to RN sample app
2. Sign in a user
3. Killed the app with swipe up
4. Open the app
5. Sign out (dispatches a Hub event an inspection of local storage shows the user is deleted)
6. Call currentAuthenticatedUser (observe there is no user)

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by elorzafe — DISMISSED** [review]

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -2155,16 +2155,23 @@ export class AuthClass {
...
 			logger.debug('failed to clear cached items');
 		}
 
 		if (this.userPool) {
-			const user = this.userPool.getCurrentUser();
-			if (user) {
-				await this.cognitoIdentitySignOut(opts, user);
+			try {
+				await this._storageSync;
+			} catch (e) {
```
Comment (HuiSF) [inline packages/auth/src/Auth.ts:2168]: Isn't the `e` already an `Error` object so `new Error()` is not needed?

File: packages/auth/src/Auth.ts [hunk 2]
```diff
@@ -2155,16 +2155,23 @@ export class AuthClass {
...
+				logger.debug('Failed to sync cache info into memory', e);
+				throw new Error(e);
+			}
+			storedCurrUser = this.userPool.getCurrentUser();
+			if (storedCurrUser) {
+				await this.cognitoIdentitySignOut(opts??{}, storedCurrUser);
```
Comment (HuiSF) [inline packages/auth/src/Auth.ts:2174]: nit
```suggestion
				await this.cognitoIdentitySignOut(opts ?? {}, storedCurrUser);
```

File: packages/auth/src/Auth.ts (hunk 1 above)
Comment (Samaritan1011001) [reply packages/auth/src/Auth.ts:2168]: good call, I used it from another place. Throwing e now like in other places again.

File: packages/auth/src/Auth.ts [hunk 3]
```diff
@@ -2155,16 +2155,23 @@ export class AuthClass {
...
-			const user = this.userPool.getCurrentUser();
-			if (user) {
-				await this.cognitoIdentitySignOut(opts, user);
+			try {
+				await this._storageSync;
+			} catch (e) {
+				logger.debug('Failed to sync cache info into memory', e);
+				throw e;
+			}
+			storedCurrUser = this.userPool.getCurrentUser();
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/auth/src/Auth.ts:2172]: If is used outside of this if block to dispatch hub event. Previously it emitted with `this.user` which was unavailable when the app booted from a killed state. But i'll revert that change to keep the fix minimal.

File: packages/auth/src/Auth.ts [hunk 4]
```diff
@@ -2155,16 +2155,23 @@ export class AuthClass {
 	 * @return - A promise resolved if success
 	 */
 	public async signOut(opts?: SignOutOpts): Promise<any> {
+		let storedCurrUser;
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/auth/src/Auth.ts:2158]: because it comes from the storage but based on the other comment, I'll simply keep what was before.

**Review by elorzafe — APPROVED** [review]

**Review by jimblanc — DISMISSED** [review]
MSTM

File: packages/auth/src/Auth.ts [hunk 5]
```diff
@@ -2154,17 +2154,23 @@ export class AuthClass {
...
+			try {
+				await this._storageSync;
+			} catch (e) {
+				logger.debug('Failed to sync cache info into memory', e);
+				throw e;
+			}
 			const user = this.userPool.getCurrentUser();
 			if (user) {
-				await this.cognitoIdentitySignOut(opts, user);
+				await this.cognitoIdentitySignOut(opts ?? {}, user);
```
Comment (ashika112) [inline packages/auth/src/Auth.ts:2173]: do you need this? since you have default val at L2157.

File: packages/auth/__tests__/auth-unit-test.ts [hunk 6]
```diff
@@ -3680,7 +3713,7 @@ describe('auth unit test', () => {
 			);
 		});
 
-		test.only('User Pools and Identity Pools', async () => {
```
Comment (jimblanc) [inline packages/auth/__tests__/auth-unit-test.ts:3683]: Oop 😅

**Review by jimblanc — APPROVED** [review]

**Review by ashika112 — APPROVED** [review]

**Pull request merged**